File: src/BreakfastProvider.Api/Grpc/BreakfastGrpcService.cs
using BreakfastProvider.Api.Storage;
using Grpc.Core;

namespace BreakfastProvider.Api.Grpc;

public class BreakfastGrpcService(
    ICosmosRepository<OrderDocument> orderRepository,
    ILogger<BreakfastGrpcService> logger) : BreakfastGrpc.BreakfastGrpcBase
{
    public override async Task<RecipeSummaryReply> GetRecipeSummary(RecipeSummaryRequest request, ServerCallContext context)
    {
        logger.LogInformation("gRPC GetRecipeSummary called for {RecipeType}", request.RecipeType);

        var reply = new RecipeSummaryReply
        {
            RecipeType = request.RecipeType,
            TotalBatches = request.RecipeType switch
            {
                "Pancakes" => 42,
                "Waffles" => 28,
                _ => 0
            },
            LastPreparedAt = DateTime.UtcNow.ToString("O")
        };

        reply.CommonIngredients.AddRange(request.RecipeType switch
        {
            "Pancakes" => ["Milk", "Flour", "Eggs"],
            "Waffles" => ["Milk", "Flour", "Eggs", "Butter"],
            _ => []
        });

        return reply;
    }

    public override async Task<OrderStatusReply> GetOrderStatus(OrderStatusRequest request, ServerCallContext context)
    {
        logger.LogInformation("gRPC GetOrderStatus called for {OrderId}", request.OrderId);

        try
        {
            var order = await orderRepository.GetByIdAsync(request.OrderId, request.OrderId, context.CancellationToken);
            return new OrderStatusReply
            {
                OrderId = order.OrderId.ToString(),
                Status = order.Status,
                CustomerName = order.CustomerName,
                ItemCount = order.Items.Count,
                CreatedAt = order.CreatedAt.ToString("O")
            };
        }
        catch (Exception)
        {
            throw new RpcException(new Status(StatusCode.NotFound, $"Order {request.OrderId} not found"));
        }
    }

    public override async Task StreamOrderUpdates(OrderStatusRequest request, IServerStreamWriter<OrderStatusReply> responseStream, ServerCallContext context)
    {
        logger.LogInformation("gRPC StreamOrderUpdates started for {OrderId}", request.OrderId);

        // Send the current status as the first message
        try
        {
            var order = await orderRepository.GetByIdAsync(request.OrderId, request.OrderId, context.CancellationToken);
            await responseStream.WriteAsync(new OrderStatusReply
            {
                OrderId = order.OrderId.ToString(),
                Status = order.Status,
                CustomerName = order.CustomerName,
                ItemCount = order.Items.Count,
                CreatedAt = order.CreatedAt.ToString("O")
            }, context.CancellationToken);
        }
        catch (Exception)
        {
            throw new RpcException(new Status(StatusCode.NotFound, $"Order {request.OrderId} not found"));
        }
    }
}

File: java/breakfast-provider-app/src/main/java/io/lemonlion/breakfast/grpc/BreakfastGrpcService.java
package io.lemonlion.breakfast.grpc;

import io.grpc.Status;
import io.grpc.stub.StreamObserver;
import io.lemonlion.breakfast.persistence.CosmosRepository;
import io.lemonlion.breakfast.storage.OrderDocument;
import java.time.Instant;
import java.util.List;
import java.util.Optional;
import net.devh.boot.grpc.server.service.GrpcService;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

/**
 * Twin of the C# {@code BreakfastGrpcService} (Grpc.AspNetCore). Exposes recipe summaries (hard-coded
 * per recipe type) and order status lookups (read from the Cosmos {@code orders} container). Registered
 * as a net.devh {@link GrpcService} so the Kronikol4J server interceptor attributes its work to the
 * calling test.
 */
@GrpcService
public class BreakfastGrpcService extends BreakfastGrpcGrpc.BreakfastGrpcImplBase {

    private static final Logger log = LoggerFactory.getLogger(BreakfastGrpcService.class);

    private final CosmosRepository<OrderDocument> orderRepository;

    public BreakfastGrpcService(CosmosRepository<OrderDocument> orderRepository) {
        this.orderRepository = orderRepository;
    }

    @Override
    public void getRecipeSummary(RecipeSummaryRequest request,
                                 StreamObserver<RecipeSummaryReply> responseObserver) {
        log.info("gRPC GetRecipeSummary called for {}", request.getRecipeType());

        int totalBatches = switch (request.getRecipeType()) {
            case "Pancakes" -> 42;
            case "Waffles" -> 28;
            default -> 0;
        };
        List<String> ingredients = switch (request.getRecipeType()) {
            case "Pancakes" -> List.of("Milk", "Flour", "Eggs");
            case "Waffles" -> List.of("Milk", "Flour", "Eggs", "Butter");
            default -> List.of();
        };

        RecipeSummaryReply reply = RecipeSummaryReply.newBuilder()
                .setRecipeType(request.getRecipeType())
                .setTotalBatches(totalBatches)
                .addAllCommonIngredients(ingredients)
                .setLastPreparedAt(Instant.now().toString())
                .build();

        responseObserver.onNext(reply);
        responseObserver.onCompleted();
    }

    @Override
    public void getOrderStatus(OrderStatusRequest request,
                               StreamObserver<OrderStatusReply> responseObserver) {
        log.info("gRPC GetOrderStatus called for {}", request.getOrderId());

        Optional<OrderStatusReply> reply = lookupOrder(request.getOrderId());
        if (reply.isEmpty()) {
            responseObserver.onError(notFound(request.getOrderId()));
            return;
        }
        responseObserver.onNext(reply.get());
        responseObserver.onCompleted();
    }

    @Override
    public void streamOrderUpdates(OrderStatusRequest request,
                                   StreamObserver<OrderStatusReply> responseObserver) {
        log.info("gRPC StreamOrderUpdates started for {}", request.getOrderId());

        Optional<OrderStatusReply> reply = lookupOrder(request.getOrderId());
        if (reply.isEmpty()) {
            responseObserver.onError(notFound(request.getOrderId()));
            return;
        }
        // Send the current status as the first (and only) message, mirroring the C# stream.
        responseObserver.onNext(reply.get());
        responseObserver.onCompleted();
    }

    private Optional<OrderStatusReply> lookupOrder(String orderId) {
        return orderRepository.findById(orderId, orderId).map(order -> OrderStatusReply.newBuilder()
                .setOrderId(order.getOrderId().toString())
                .setStatus(order.getStatus())
                .setCustomerName(order.getCustomerName())
                .setItemCount(order.getItems().size())
                .setCreatedAt(order.getCreatedAt().toString())
                .build());
    }

    private static io.grpc.StatusRuntimeException notFound(String orderId) {
        return Status.NOT_FOUND.withDescription("Order " + orderId + " not found").asRuntimeException();
    }
}
